api/envoy/extensions/load_balancing_policies/deterministic_aperture/v3/deterministic_aperture.proto

```diff
@@ -0,0 +1,61 @@
+syntax = "proto3";
+
+package envoy.extensions.load_balancing_policies.deterministic_aperture.v3;
+
+import "google/protobuf/wrappers.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.load_balancing_policies.deterministic_aperture.v3";
+option java_outer_classname = "DeterministicApertureProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/deterministic_aperture/v3;deterministic_aperturev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Deterministic Aperture Load Balancing Policy]
+// [#extension: envoy.load_balancing_policies]
+
+// Deterministic Aperture load balancing is designed to conserve connections between Envoy and
+// backend instances. This is achieved by dividing the backends uniformly among participating peers.
+// This way, we avoid a complete mesh of connections between Envoy and backends. The participating
+// Envoy peers are placed on a ring that is conceptually laid on top of the Ring Hash of backends
+// to determine their overlaps. By using the ring overlaps as a mechanism to divide the backends,
+// the backends get divided as a fraction of their overlaps with each peer. The algorithm uses a
+// uniform random distribution to select a backend from the range of overlapping backends. This
+// along with P2C ensures uniform load distribution.
+// [#next-free-field: 7]
+message DeterministicApertureLbConfig {
```

Comment:
This config still needs a field for the child policy config, or else there's no way to tell it what policy to actually use to pick from amongst the selected subset of endpoints.  I think you need to add something like this:

```
config.cluster.v3.LoadBalancingPolicy child_policy = 7
    [(validate.rules).message = {required: true}];
```

Reply:
@markdroth Thanks for looking. Not sure what you mean. The DeterministicAperture load balancer policy is itself a `LbPolicy` of type `LOAD_BALANCING_POLICY_CONFIG`.  The integration test in the PR shows the usage.

Reply:
@markdroth The `policy` here is at the same level as other load balancers in Envoy. So `deterministic aperture` is able to `choose` host from the subset of endpoints. This subset is derived by the `deterministic aperture` load balancer by overlapping the `ring` of endpoints with the `ring` of participating Envoys. So in short the load balancer here is self contained (like others - aka Maglev, Ring etc). Please let me know if that makes sense.

Reply:
I understand that this is intended to be used as a top-level LB policy.  But that doesn't mean that it can't delegate to another policy for the actual pick from among the filtered subset of hosts.

What is the actual picking algorithm used here?  Is there any reason that that algorithm needs to be tied to the subsetting behavior?  If not, it seems more flexible to delegate to a child policy.

Reply:
The actual picking algorithm is described in the `load_balancer.h` header. Copying the comments here:
```
Implementation:
 *      - Uses a `Hash Ring` for the endpoints.
 *      - Envoy peers are placed on a ring that is conceptually laid on top of the Ring Hash of
 * backends to determine their overlaps.
 *      - By using the ring overlaps as a mechanism to divide the
 * backends, the backends get divided as a fraction of their overlaps with each peer.
 *      - The algorithm uses a uniform random distribution to select a backend from the range of
 * overlapping backends. This along with P2C ensures uniform load distribution.
 */


```

The implementation is in `LoadBalancer::chooseHost` method. The implementation was already reviewed by @zuercher .

Reply:
Thanks, I understand the algorithm now.  But you didn't respond to my question about why that algorithm needs to be hard-coded along with the subsetting behavior.

Wouldn't the behavior be strictly more flexible if we split it up into two policies, so that users could choose whether to use the subsetting behavior independently from choosing which pick algorithm to use?  Couldn't there be cases where someone might want the uniform random distribution picking algorithm you describe without the subsetting?  And couldn't there be cases where someone might want the subsetting behavior with a picking policy other than this uniform random distribution algorithm?

Unless there is some reason that these two steps cannot work independently of each other, it seems unnecessarily inflexible to forcibly tie them together.  I don't understand why we wouldn't want to split them up.

Reply:
That is a possibility but currently not in the scope. I might add that flexibility in the future but for now, simplicity is the goal. As you said, we could imagine in the future something that allows `pick` strategy as an option. Currently it is strictly p2c.

Reply:
I understand that you don't currently need the flexibility of independently selecting the subsetting behavior from the picking behavior.  But implementing this in a monolithic way now means that when we do need the flexibility in the future, Envoy will be stuck supporting this as both a monolithic policy and as separate and composable policies.  In contrast, if we just implement this as separate and composable policies from the get-go, we can address both use-cases without taking on that tech debt.

I'm not asking you to change the actual algorithm here; the behavior will still match what is described in that twitter blog post.  All I'm asking is that we to do this in a cleaner way so that we don't take on that tech debt.  I don't think this should actually be that much more work.

Reply:
I could add an Enum like :

enum PickPolicy {
   P2C
}
So that we can choose the `pick` policy. Or I could add more complex structure/abstraction. Please let me know.
In reality,  I think we wont be expanding on the P2C policy mostly because this algorithm is an optimization of P2C.

Reply:
I think that using an enum and adding a more complex abstraction basically amounts to introducing a new custom extension API within the subsetting policy, which means that anyone who wants to plugin their own picking algorithm would have to write a new plugin for that extension point, and it would not be usable outside of the subsetting policy.  And the new extension API would basically have the same functionality that the existing LB policy API does, so it's not clear to me why it would be beneficial to use a different extension API here.

I think we should just use the existing LB policy API.  Writing a custom extension API seems like more work, more duplication of effort, and ultimately gives us less flexibility.

Reply:
Just to make sure I understand - DeterministicAperture would be a LB policy which will host another LB policy ? Since the policy here is the picking algorithm, I was hoping that that can be encapsulated via an `Enum`.  Having the picking algorithm as a policy is also ok but i am worried it would imply a separate `typed_config` , no? Please let me know. Happy to add what works best.

Reply:
Yes, it will be a separate typed config.  Specifically, as I mentioned in my original comment, I think we want to add the following field here:

```
config.cluster.v3.LoadBalancingPolicy child_policy = 7
    [(validate.rules).message = {required: true}];
```

The typed config is inside the `LoadBalancingPolicy` message.

I think the key point here is that the top-level policy should do the subsetting behavior, and it should then delegate to a child policy to pick from among the selected subset of hosts.  I don't really care that much about the names of the two policies; if you'd rather use the name "deterministic_aperature" for the lower-level picking policy instead of the top-level subsetting policy, that's fine.

Reply:
Got it. Had a question -  how do I represent "P2C" as the default policy? Do i create a factory for P2C and a separate `typed config` for it. Then i can iterate through the `policies` list in the `child_policy` in the load balancer constructor?

Reply:
Yes, I think you should create a separate factory and a separate config proto for the child policy.

I don't think the child policy should default to any particular policy; I think it makes more sense to have the config explicitly specify the child policy.

Reply:
I started refactoring the code but I am finding it a bit convoluted to fit the concept of `picking` algorithm into the LoadBalancerPolicy abstraction. The problem is that there is currently `TypedLoadBalancerFactory` which is meant to create load balancers defined by `LoadBalancingPolicy`.  But the create method for this interface has parameters that are good fit for a policies who need priority set etc. All we need in the case of the deterministic aperture is a pick algorithm. In the case of deterministic aperture, the endpoints and priority sets etc are already consumed by parent Ring to create the subsets. The picking algorithm would just need the `Ring` to get the subset. This forces me to  to create a new factory for creating the child policy that `create` a child given a Ring. 
Not sure if this is what you had in mind but i am finding this a bit too complex just to achieve which piking algorithm to use.  
Also, I dont think we can use existing load balancing policies (such as Maglev or RinhHash) as child policy since they  would need their own Ring etc.

Reply:
Just to clarify my proposal: Here is a `ChildPolicy` definition: 
```
// [#protodoc-title: Deterministic Aperture Child Policy]
message ChildPolicy {
    reserved 1,2,3;

    reserved "config", "name", "typed_config";

    core.v3.TypedExtensionConfig typed_extension_config = 4;
}
```
And the `ChildPolicy` interface and factory:
```
class ChildPolicy {
 public:
  virtual ~ChildPolicy() = default;
  virtual Upstream::HostConstSharedPtr chooseHost() PURE;
};

using std::unique_ptr<ChildPolicy> ChildPolicyPtr;

/**
 * Factory config for deterministic aperture load balancer's child policies.
 */
class TypedChildPolicyFactory : public Config::TypedFactory {
public:
  ~TypedChildPolicyFactory() override = default;

  std::string category() const override {
    return "envoy.load_balancing_policies.deterministic_aperture.child_policy";
  }

  /**
   * @return LoadBalancerPtr a child load balancer policy implementation.
   */
  virtual ChildPolicyPtr create(Ring* ring, Stats::ScopeSharedPtr) PURE;
};
```

`P2C` specific implementation:

```
class P2CPolicy : public ChildPolicy {
public:
  P2CPolicy(Ring*, Stats::ScopeSharedPtr);

  Upstream::HostConstSharedPtr chooseHost() override;

private:
  
  /*
   * Pick an index as part of the `p2c` algorithm.
   * Uses uniform random distribution within to pick a random index in the peer's `offset` and
   * `width`.
   */
  size_t pick() const;

  /*
   * Pick another index in the peer's `offset` and `width` range.
   * @param first Index that was already picked. The new pick cannot overlap the first pick's
   * region.
   */
  size_t pickSecond(size_t first) const;

  /*
   * Picks two indexes as per the `p2c` algorithm.
   */
  std::pair<size_t, size_t> pick2() const;

  Ring* parent_;
 
};
```

Reply:
I understand what you are proposing, but I don't think it makes any sense to create a new interface here.  The whole point of an LB policy is to provide the endpoint picking behavior across a set of endpoints, so creating a new interface to do the same thing makes no sense.  That would add needless complexity and prevent people from using any of the existing policies as child policies.

To be clear, what I am suggesting here is that the P2C child policy should create its own ring rather than reusing the ring from the parent subsetting policy.  There should not be any need to share any state between the parent and child policy except for the information that can be passed through the existing LB policy API.

Reply:
I am missing something here. The P2C Policy with  its own Ring is the current state of the PR.  The current PR exposes the P2C Policy + Ring as a load balancing policy which can be used as shown in the integration test.

Reply:
What I am saying is that there should be two different rings here: one in the parent policy that does subsetting, and one in the child policy that picks the endpoint from the selected subset.  You should not need to pass the ring through the LB policy API when creating the child.

Reply:
The child (P2C) would need access to the parent Ring (to ask what are my subsets). I am not sure in the case of P2C, if it needs a Ring for itself since all it needs is the section of the Ring it can pick from. The parent Ring provides API for child to ask `give me the ring subsection(subset)` so that I can apply my policy. Some  other policy might decide to create its own Ring with the subset information but that depends on that policy.

Reply:
Let's step back for a moment here.  Let's set aside the details of the specific algorithm used in the two policies and instead talk about what each of the two policies is responsible for and how they interact with each other.

The LB policy API basically gives the policy a set of hosts and then asks it to pick from that set of hosts for each request.  We want both the parent and the child policy to use that same API.

The goal of the parent policy is to decide what subset of hosts the child should pick from.  In effect, the parent policy does not actually do any picking; all it does is filter the set of hosts that it is given so that it passes only a subset of those hosts to the child policy.

The goal of the child policy is to pick from the set of hosts that it is given by its parent.  It should not matter to the child whether its parent is Envoy (i.e., it is the top-level policy) or whether its parent is another LB policy that did subsetting before delegating to it; either way, it is just responsible for picking from amongst the hosts it's been given.

I think the important properties here are:
1. Both of these policies are implemented using the existing LB policy API, not creating some new API to do the same thing.
2. There should be no communication between the two policies that cannot be expressed via the existing LB policy API.

So getting back to this specific situation, I don't think we want an API for the child to refer back to the ring in the parent policy, because that would tie the child to this parent; the child would no longer be able to be used by itself without the parent.  Instead, I think we want the parent policy to build the ring, decide which subset to use, and then construct a list of hosts in that subset.  The parent policy will then pass that list of hosts to the child in the same format as would be expected by any LB policy.

If the child policy doesn't actually need a ring for its algorithm, then it doesn't need to construct a ring.  It can just do a uniform random distribution pick from among the set of hosts that are passed to it.